improve broadcast documentation (#871)
cylc/cylc-doc: src/user-guide/running-workflows/cylc-broadcast.rst

.. _cylc-broadcast:

Cylc Broadcast
==============

Cylc "Broadcasts" allow us to override task's :cylc:conf:`[runtime]`
settings within a running workflow.

Broadcasts can target specific cycles, families or tasks.

Broadcasts can be helpful for:

* Quickly developing tasks without having to
  :ref:`edit and reload the workflow configuration <interventions.edit-the-workflow-configuration>`.
* Sending small amounts of data from task to another (e.g. file paths).

> * Orchestrating production workflows from an external system.

Author: Fine:

```suggestion
* Reconfiguring production workflows while they are running.
```

Once issued, broadcasts last for the life of the workflow and will persist if
the workflow is shutdown and restarted (unless manually "cleared").

However, broadcasts which target specific cycles will eventually be
:ref:`expired <user_guide.broadcast.expiry>` when no longer needed.

.. seealso::

   :ref:`broadcast-tutorial`

Issuing Broadcasts
------------------

CLI
^^^

Some examples of issuing broadcasts using the ``cylc broacast`` CLI command:

.. code-block:: bash

   # set or update the environment variable "ANSWER" for all tasks
   cylc broadcast myworkflow -s '[environment]ANSWER=42'

   # amend the directives for all tasks in the cycle "2000"
   cylc broadcast myworkflow -p 2000 -s '[directives]--memory=2GB'

   # change the platform all tasks in the family "FOO" will submit on
   cylc broadcast myworkflow -n FOO -s 'platform=my-hpc'

The ``cylc broadcast`` command can be run from within tasks as a means to
communicate small amounts of data back to the scheduler for subsequent tasks to
use. Note that this will not work for remote task platforms which have been
configured to use polling -

> :cylc:conf:`global.cylc[platforms][<platform name>]communication method = poll`.

Author: If TCP comms are possible, but polling is configured, it still won't work. So it's more the use of polling than the lack of TCP.

Note, broadcasts will also work over SSH comms.

.. _cylc-broadcast:

Cylc Broadcast
==============

> Cylc "Broadcasts" allow us to override task's :cylc:conf:`[runtime]`

Reviewer: (Replacing original unresolved suggestion after GitHub UI screwed up an edit).

"override task's runtime" isn't good grammar - it should be "*override tasks' runtime*" (possession by tasks, plural) or "*override task runtimes*" or possibly even "*override task runtime*" (taking "task runtime" as a compound noun).

Maybe a bit more explicit for clarity though:

```suggestion
Cylc "Broadcasts" allow us to override the :cylc:conf:`[runtime]` settings of tasks.
```

> settings within a running workflow.

Reviewer: ```suggestion
```

Reviewer: (Kinda my fault - I push the change and merge),

   cylc broadcast myworkflow -n FOO -s 'platform=my-hpc'

The ``cylc broadcast`` command can be run from within tasks as a means to
communicate small amounts of data back to the scheduler for subsequent tasks to
use. Note that this will not work for remote task platforms which have been
configured to use polling -
:cylc:conf:`global.cylc[platforms][<platform name>]communication method = poll`.

For more information, run ``cylc broadcast --help``.

GUI
^^^

Broadcasts can also be issued from the GUI in a similar way by choosing
"Broadcast" from the task menu.

Additionally, the GUI provides a utility called "Edit Runtime" which loads
the tasks configuration into a form. Any changes you make are then broadcasted
to the task:

.. image:: ../interventions/edit-a-tasks-configuration.gui.gif
   :width: 75%
   :align: center

|

.. _user_guide.broadcast.expiry:

Expiry
------

Broadcasts which target specific cycles will eventually expire (i.e. be
deleted) as the workflow moves on, to avoid gradual accumulation
(note broadcasts are persisted when the workflow restarts).

Expiry Point
^^^^^^^^^^^^

Broadcasts expire once are they are no longer required by upcoming tasks.

> The exact point at which a broadcast is expired depends on two things:

Reviewer 2: I can count to 2 (sometimes).

```suggestion
The exact point at which a broadcast is expired depends on:
```
